docs/proposals/20230117-cluster-level-identity.md

```diff
@@ -0,0 +1,186 @@
+
+```go
+// OCIClusterSpec defines the desired state of OciCluster
+type OCIClusterSpec struct {
+  // IdentityRef is a reference to an identity(principal) to be used when reconciling this cluster
+  // +optional
+  IdentityRef *corev1.ObjectReference `json:"identityRef,omitempty"`
+}
+```
+The following custom resources will be added.
+
+```go
+
+type PrincipalType string
+
+const (
+    // UserPrincipal represents a user princpal.
+    UserPrincipal PrincipalType = "UserPrincipal"
+)
+
+// OCIClusterIdentitySpec defines the parameters that are used to create an OCIIdentity.
+type OCIClusterIdentitySpec struct {
+  // Type is the type of OCI Principal used.
+  // UserPrincipal is the only supported value
+  Type PrincipalType `json:"type"`
+  
+  // PrincipalSecret is a secret reference which contains the authentication credentials for the principal.
+  // +optional
+  PrincipalSecret corev1.SecretReference `json:"clientSecret,omitempty"`
+
+  // AllowedNamespaces is used to identify the namespaces the clusters are allowed to use the identity from.
+  // Namespaces can be selected either using an array of namespaces or with label selector.
+  // An empty allowedNamespaces object indicates that OCIClusters can use this identity from any namespace.
+  // If this object is nil, no namespaces will be allowed (default behaviour, if this field is not provided)
+  // A namespace should be either in the NamespaceList or match with Selector to use the identity.
+  //
+  // +optional
+  // +nullable
+  AllowedNamespaces *AllowedNamespaces `json:"allowedNamespaces"`
+}
+
+// AllowedNamespaces defines the namespaces the clusters are allowed to use the identity from
+// NamespaceList takes precedence over the Selector.
+type AllowedNamespaces struct {
+  // A nil or empty list indicates that OCICluster cannot use the identity from any namespace.
+  //
+  // +optional
+  // +nullable
+  NamespaceList []string `json:"list"`
+  
+  // Selector is a selector of namespaces that OCICluster can
+  // use this Identity from. This is a standard Kubernetes LabelSelector,
+  // a label query over a set of resources. The result of matchLabels and
+  // matchExpressions are ANDed.
+  //
+  // A nil or empty selector indicates that OCICluster cannot use this
+  // OCIClusterIdentity from any namespace.
+  // +optional
+  Selector *metav1.LabelSelector `json:"selector"`
+}
+
+// OCIClusterIdentityStatus defines the observed state of OCIClusterIdentity.
+type OCIClusterIdentityStatus struct {
+  // Conditions defines current service state of the OCIClusterIdentity.
+  // +optional
+  Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+}
+
+type OCIClusterIdentity struct {
+  metav1.TypeMeta   `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec   OCIClusterIdentitySpec   `json:"spec,omitempty"`
+  Status OCIClusterIdentityStatus `json:"status,omitempty"`
+}
+
+```
+
+
+### Example Cluster template
```

Comment:
Is the behavior the same when using clusterctl?

Reply:
Can you expand on this a little bit? Sorry could not understand what you meant by same behavior when using clusterctl